Upgrade dto generator mojo to support LSP4J 0.4.0

### Description

When trying to use newer (e.g. pre-release 0.4.0) LSP4J the generator fails like below. I need a newer version to support #5508.

[ERROR] Failed to execute goal
org.eclipse.che.core:che-core-api-languageserver-maven-plugin:6.0.0-M5-SNAPSHOT:generate
(default) on project che-plugin-languageserver-ide: A type
incompatibility occurred while executing
org.eclipse.che.core:che-core-api-languageserver-maven-plugin:6.0.0-M5-SNAPSHOT:generate:
java.lang.Class cannot be cast to java.lang.reflect.ParameterizedType
[ERROR] -----------------------------------------------------
[ERROR] realm =
plugin>org.eclipse.che.core:che-core-api-languageserver-maven-plugin:6.0.0-M5-SNAPSHOT
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
[ERROR] urls[0] =
file:/home/jonah/.m2/repository/org/eclipse/che/core/che-core-api-languageserver-maven-plugin/6.0.0-M5-SNAPSHOT/che-core-api-languageserver-maven-plugin-6.0.0-M5-SNAPSHOT.jar
[ERROR] urls[1] =
file:/home/jonah/.m2/repository/org/eclipse/che/core/che-core-api-languageserver-shared/6.0.0-M5-SNAPSHOT/che-core-api-languageserver-shared-6.0.0-M5-SNAPSHOT.jar
[ERROR] urls[2] =
file:/home/jonah/.m2/repository/org/eclipse/lsp4j/org.eclipse.lsp4j/0.4.0-SNAPSHOT/org.eclipse.lsp4j-0.4.0-SNAPSHOT.jar
[ERROR] urls[3] =
file:/home/jonah/.m2/repository/org/eclipse/lsp4j/org.eclipse.lsp4j.generator/0.4.0-SNAPSHOT/org.eclipse.lsp4j.generator-0.4.0-SNAPSHOT.jar
[ERROR] urls[4] =
file:/home/jonah/.m2/repository/org/eclipse/xtend/org.eclipse.xtend.lib/2.14.0-SNAPSHOT/org.eclipse.xtend.lib-2.14.0-SNAPSHOT.jar
[ERROR] urls[5] =
file:/home/jonah/.m2/repository/org/eclipse/xtext/org.eclipse.xtext.xbase.lib/2.10.0/org.eclipse.xtext.xbase.lib-2.10.0.jar
[ERROR] urls[6] =
file:/home/jonah/.m2/repository/org/eclipse/xtend/org.eclipse.xtend.lib.macro/2.14.0-SNAPSHOT/org.eclipse.xtend.lib.macro-2.14.0-SNAPSHOT.jar
[ERROR] urls[7] =
file:/home/jonah/.m2/repository/com/google/guava/guava/20.0/guava-20.0.jar
[ERROR] urls[8] =
file:/home/jonah/.m2/repository/org/codehaus/plexus/plexus-utils/3.0.20/plexus-utils-3.0.20.jar
[ERROR] urls[9] =
file:/home/jonah/.m2/repository/javax/enterprise/cdi-api/1.0/cdi-api-1.0.jar
[ERROR] urls[10] =
file:/home/jonah/.m2/repository/org/eclipse/sisu/org.eclipse.sisu.inject/0.3.0/org.eclipse.sisu.inject-0.3.0.jar
[ERROR] urls[11] =
file:/home/jonah/.m2/repository/org/codehaus/plexus/plexus-component-annotations/1.5.5/plexus-component-annotations-1.5.5.jar
[ERROR] urls[12] =
file:/home/jonah/.m2/repository/org/eclipse/lsp4j/org.eclipse.lsp4j.jsonrpc/0.4.0-SNAPSHOT/org.eclipse.lsp4j.jsonrpc-0.4.0-SNAPSHOT.jar
[ERROR] urls[13] =
file:/home/jonah/.m2/repository/com/google/code/gson/gson/2.8.2/gson-2.8.2.jar
[ERROR] urls[14] =
file:/home/jonah/.m2/repository/org/reflections/reflections/0.9.9/reflections-0.9.9.jar
[ERROR] urls[15] =
file:/home/jonah/.m2/repository/org/javassist/javassist/3.22.0-GA/javassist-3.22.0-GA.jar
[ERROR] Number of foreign imports: 1
[ERROR] import: Entry[import  from realm
ClassRealm[project>org.eclipse.che.plugin:che-plugin-languageserver-ide:6.0.0-M5-SNAPSHOT,
parent: ClassRealm[maven.api, parent: null]]]
[ERROR]
[ERROR] -----------------------------------------------------
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with
the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions,
please read the following articles:
[ERROR] [Help 1]
http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException

--- a/plugins/plugin-languageserver/che-plugin-languageserver-ide/pom.xml
+++ b/plugins/plugin-languageserver/che-plugin-languageserver-ide/pom.xml
@@ -193,6 +193,7 @@
                     <excludes>
                         <exclude>**/org/eclipse/che/plugin/languageserver/ide/editor/codeassist/LatestCompletionResult.java</exclude>
                         <exclude>**/Either.java</exclude>
+                        <exclude>**/Either3.java</exclude>
                     </excludes>
                 </configuration>
             </plugin>

--- a/plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/resources/org/eclipse/lsp4j/jsonrpc/messages/Either.java
+++ b/plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/resources/org/eclipse/lsp4j/jsonrpc/messages/Either.java
@@ -1,14 +1,27 @@
+/*******************************************************************************
+ * Copyright (c) 2016 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
 package org.eclipse.lsp4j.jsonrpc.messages;
+
 
 import org.eclipse.lsp4j.jsonrpc.validation.NonNull;
 
 /**
  * An either type maps union types in protocol specifications.
- *
- * @param <L>
- * @param <R>
  */
 public class Either<L, R> {
+
+    public static <L, R> Either<L, R> forLeft(@NonNull L left) {
+        return new Either<L, R>(left, null);
+    }
+
+    public static <L, R> Either<L, R> forRight(@NonNull R right) {
+        return new Either<L, R>(null, right);
+    }
 
     private final L left;
     private final R right;
@@ -17,14 +30,6 @@ public class Either<L, R> {
         super();
         this.left = left;
         this.right = right;
-    }
-
-    public static <L, R> Either<L, R> forLeft(@NonNull L left) {
-        return new Either<L, R>(left, null);
-    }
-
-    public static <L, R> Either<L, R> forRight(@NonNull R right) {
-        return new Either<L, R>(null, right);
     }
 
     public L getLeft() {
@@ -41,6 +46,25 @@ public class Either<L, R> {
 
     public boolean isRight() {
         return right != null;
+    }
+    
+    @Override
+    public boolean equals(Object obj) {
+        if (obj instanceof Either<?, ?>) {
+            Either<?, ?> other = (Either<?, ?>) obj;
+            return this.left != null && other.left != null && this.left.equals(other.left)
+                    || this.right != null && other.right != null && this.right.equals(other.right);
+        }
+        return false;
+    }
+    
+    @Override
+    public int hashCode() {
+        if (this.left != null)
+            return this.left.hashCode();
+        if (this.right != null)
+            return this.right.hashCode();
+        return 0;
     }
 
     public String toString() {

--- a/plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/resources/org/eclipse/lsp4j/jsonrpc/messages/Either3.java
+++ b/plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/resources/org/eclipse/lsp4j/jsonrpc/messages/Either3.java
@@ -1,0 +1,82 @@
+/*******************************************************************************
+ * Copyright (c) 2017 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.lsp4j.jsonrpc.messages;
+
+import org.eclipse.lsp4j.jsonrpc.validation.NonNull;
+
+/**
+ * Union type for three types.
+ */
+public class Either3<T1, T2, T3> extends Either<T1, Either<T2, T3>> {
+    
+    public static <T1, T2, T3> Either3<T1, T2, T3> forFirst(@NonNull T1 first) {
+        return new Either3<T1, T2, T3>(first, null);
+    }
+
+    public static <T1, T2, T3> Either3<T1, T2, T3> forSecond(@NonNull T2 second) {
+        return new Either3<T1, T2, T3>(null, new Either<T2, T3>(second, null));
+    }
+    
+    public static <T1, T2, T3> Either3<T1, T2, T3> forThird(@NonNull T3 third) {
+        return new Either3<T1, T2, T3>(null, new Either<T2, T3>(null, third));
+    }
+    
+    public static <T1, T2, T3> Either3<T1, T2, T3> forLeft3(@NonNull T1 first) {
+        return new Either3<T1, T2, T3>(first, null);
+    }
+    
+    public static <T1, T2, T3> Either3<T1, T2, T3> forRight3(@NonNull Either<T2, T3> right) {
+        return new Either3<T1, T2, T3>(null, right);
+    }
+
+    protected Either3(T1 left, Either<T2, T3> right) {
+        super(left, right);
+    }
+    
+    public T1 getFirst() {
+        return getLeft();
+    }
+    
+    public T2 getSecond() {
+        Either<T2, T3> right = getRight();
+        if (right == null)
+            return null;
+        else
+            return right.getLeft();
+    }
+    
+    public T3 getThird() {
+        Either<T2, T3> right = getRight();
+        if (right == null)
+            return null;
+        else
+            return right.getRight();
+    }
+    
+    public boolean isFirst() {
+        return isLeft();
+    }
+    
+    public boolean isSecond() {
+        return isRight() && getRight().isLeft();
+    }
+    
+    public boolean isThird() {
+        return isRight() && getRight().isRight();
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder("Either3 [").append("\n");
+        builder.append("  first = ").append(getFirst()).append("\n");
+        builder.append("  second = ").append(getSecond()).append("\n");
+        builder.append("  third = ").append(getThird()).append("\n");
+        return builder.append("]").toString();
+    }
+
+}

--- a/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/DtoGenerator.java
+++ b/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/DtoGenerator.java
@@ -161,6 +161,7 @@ public abstract class DtoGenerator {
       out.println("import java.util.ArrayList;");
       out.println("import java.util.HashMap;");
       out.println("import org.eclipse.lsp4j.jsonrpc.messages.Either;");
+      out.println("import org.eclipse.lsp4j.jsonrpc.messages.Either3;");
       out.println("import org.eclipse.che.api.languageserver.util.EitherUtil;");
       out.println("import org.eclipse.che.api.languageserver.util.JsonUtil;");
       out.println("import org.eclipse.che.api.languageserver.shared.util.JsonDecision;");

--- a/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/EitherUtil.java
+++ b/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/EitherUtil.java
@@ -29,6 +29,18 @@ public class EitherUtil {
     return getDisjointType(type, 1);
   }
 
+  static Type getFirstDisjointType(Type type) {
+    return getDisjointType(type, 0);
+  }
+
+  static Type getSecondDisjointType(Type type) {
+    return getDisjointType(type, 1);
+  }
+
+  static Type getThirdDisjointType(Type type) {
+    return getDisjointType(type, 2);
+  }
+
   private static Type getDisjointType(Type type, int index) {
     if (type instanceof ParameterizedType) {
       final ParameterizedType parameterizedType = (ParameterizedType) type;

--- a/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/FromJsonGenerator.java
+++ b/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/FromJsonGenerator.java
@@ -23,6 +23,7 @@ import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.eclipse.lsp4j.jsonrpc.messages.Either3;
 
 /**
  * This class generates property conversion code from json properties to dto fields.
@@ -70,6 +71,8 @@ public class FromJsonGenerator extends ConversionGenerator {
       generateListConversion(indent, out, varName, valueAccess, paramType);
     } else if (Map.class.isAssignableFrom(getRawClass(paramType))) {
       generateMapConversion(indent, out, varName, valueAccess, paramType);
+    } else if (Either3.class.isAssignableFrom(getRawClass(paramType))) {
+      generateEither3Conversion(indent, out, varName, valueAccess, paramType);
     } else if (Either.class.isAssignableFrom(getRawClass(paramType))) {
       generateEitherConversion(indent, out, varName, valueAccess, paramType);
     } else {
@@ -85,6 +88,9 @@ public class FromJsonGenerator extends ConversionGenerator {
 
   private void generateMapConversion(
       String indent, PrintWriter out, String varName, String jsonValName, Type paramType) {
+    if (!(paramType instanceof ParameterizedType)) {
+      paramType = ((Class<?>) paramType).getGenericSuperclass();
+    }
     ParameterizedType genericType = (ParameterizedType) paramType;
     Type containedType = genericType.getActualTypeArguments()[1];
     String objectName = varName + "o";
@@ -157,6 +163,50 @@ public class FromJsonGenerator extends ConversionGenerator {
     return "JsonDecision.OBJECT";
   }
 
+  private void generateEither3Conversion(
+      String indent, PrintWriter out, String varName, String valueAccess, Type paramType) {
+    String innerName = varName + "e";
+
+    out.println(indent + String.format("%1$s %2$s;", paramType.getTypeName(), varName));
+    String firstDecisions = getJsonDecisions(EitherUtil.getFirstDisjointType(paramType));
+    String secondDecisions = getJsonDecisions(EitherUtil.getSecondDisjointType(paramType));
+    out.println(
+        indent
+            + String.format("if (EitherUtil.matches(%1$s, %2$s)) {", valueAccess, firstDecisions));
+
+    generateFromJson(
+        indent + INDENT, out, innerName, valueAccess, EitherUtil.getFirstDisjointType(paramType));
+
+    out.println(
+        indent + INDENT + String.format("%1$s= Either3.forFirst(%2$s);", varName, innerName));
+    out.println(
+        indent
+            + String.format(
+                "} else if (EitherUtil.matches(%1$s, %2$s)) {", valueAccess, secondDecisions));
+
+    generateFromJson(
+        indent + INDENT, out, innerName, valueAccess, EitherUtil.getSecondDisjointType(paramType));
+
+    out.println(
+        indent + INDENT + String.format("%1$s= Either3.forSecond(%2$s);", varName, innerName));
+    out.println(indent + "} else  {");
+
+    generateFromJson(
+        indent + INDENT, out, innerName, valueAccess, EitherUtil.getThirdDisjointType(paramType));
+    out.println(
+        indent + INDENT + String.format("%1$s= Either3.forThird(%2$s);", varName, innerName));
+    out.println(indent + "}");
+  }
+
+  private String getJsonDecisions(Type firstDisjointType) {
+    String decisionNames =
+        EitherUtil.getAllDisjoinTypes(firstDisjointType)
+            .stream()
+            .map(t -> getJsonDecision(getRawClass(t)))
+            .collect(Collectors.joining(","));
+    return String.format("new JsonDecision[] { %1$s }", decisionNames);
+  }
+
   private void generateEitherConversion(
       String indent, PrintWriter out, String varName, String valueAccess, Type paramType) {
     String innerName = varName + "e";
@@ -220,6 +270,8 @@ public class FromJsonGenerator extends ConversionGenerator {
   private Object primitiveCast(Class<? extends Number> t) {
     if (t == Integer.class) {
       return "int";
+    } else if (t == Number.class) {
+      return "Number";
     } else {
       return t.getSimpleName().toLowerCase();
     }

--- a/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/ToDtoGenerator.java
+++ b/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/ToDtoGenerator.java
@@ -21,6 +21,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.eclipse.lsp4j.jsonrpc.messages.Either3;
 
 /**
  * This class generates property conversions from regular lsp4j classes to dto classes. Used to
@@ -90,6 +91,8 @@ public class ToDtoGenerator extends ConversionGenerator {
       generateListConversion(indent + INDENT, out, varName, valueAccess, paramType);
     } else if (Map.class.isAssignableFrom(rawClass)) {
       generateMapConversion(indent + INDENT, out, varName, valueAccess, paramType);
+    } else if (Either3.class.isAssignableFrom(getRawClass(paramType))) {
+      generateEither3Conversion(indent + INDENT, out, varName, valueAccess, paramType);
     } else if (Either.class.isAssignableFrom(rawClass)) {
       generateEitherConversion(indent + INDENT, out, varName, valueAccess, paramType);
     } else {
@@ -100,6 +103,41 @@ public class ToDtoGenerator extends ConversionGenerator {
               varName,
               dtoValueExpression(rawClass, valueAccess)));
     }
+  }
+
+  private void generateEither3Conversion(
+      String indent, PrintWriter out, String varName, String valueAccess, Type paramType) {
+    String innerName = varName + "e";
+
+    out.println(indent + String.format("%1$s %2$s;", paramType.getTypeName(), varName));
+    out.println(indent + String.format("if (%1$s.isFirst()) {", valueAccess));
+    generateToDto(
+        indent + INDENT,
+        out,
+        innerName,
+        valueAccess + ".getFirst()",
+        EitherUtil.getFirstDisjointType(paramType));
+    out.println(
+        indent + INDENT + String.format("%1$s= Either3.forFirst(%2$s);", varName, innerName));
+    out.println(indent + String.format("} else if (%1$s.isSecond()) {", valueAccess));
+    generateToDto(
+        indent + INDENT,
+        out,
+        innerName,
+        valueAccess + ".getSecond()",
+        EitherUtil.getSecondDisjointType(paramType));
+    out.println(
+        indent + INDENT + String.format("%1$s= Either3.forSecond(%2$s);", varName, innerName));
+    out.println(indent + "} else  {");
+    generateToDto(
+        indent + INDENT,
+        out,
+        innerName,
+        valueAccess + ".getThird()",
+        EitherUtil.getThirdDisjointType(paramType));
+    out.println(
+        indent + INDENT + String.format("%1$s= Either3.forThird(%2$s);", varName, innerName));
+    out.println(indent + "}");
   }
 
   private void generateEitherConversion(
@@ -129,6 +167,9 @@ public class ToDtoGenerator extends ConversionGenerator {
 
   private void generateMapConversion(
       String indent, PrintWriter out, String varName, String valueAccess, Type paramType) {
+    if (!(paramType instanceof ParameterizedType)) {
+      paramType = ((Class<?>) paramType).getGenericSuperclass();
+    }
     ParameterizedType genericType = (ParameterizedType) paramType;
     Type containedType = genericType.getActualTypeArguments()[1];
     Type valueType = genericType.getActualTypeArguments()[1];

--- a/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/ToJsonGenerator.java
+++ b/wsagent/che-core-api-languageserver-maven-plugin/src/main/java/org/eclipse/che/api/languageserver/generator/ToJsonGenerator.java
@@ -20,6 +20,7 @@ import java.lang.reflect.Type;
 import java.util.List;
 import java.util.Map;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.eclipse.lsp4j.jsonrpc.messages.Either3;
 
 /**
  * This class generates code to convert dto object fields to json properties.
@@ -64,6 +65,8 @@ public class ToJsonGenerator extends ConversionGenerator {
       new ListConverter(varName, valueAccess, paramType).generateListConversion(indent, out);
     } else if (Map.class.isAssignableFrom(getRawClass(paramType))) {
       generateMapConversion(indent, out, varName, valueAccess, paramType);
+    } else if (Either3.class.isAssignableFrom(getRawClass(paramType))) {
+      generateEither3Conversion(indent, out, varName, valueAccess, paramType);
     } else if (Either.class.isAssignableFrom(getRawClass(paramType))) {
       generateEitherConversion(indent, out, varName, valueAccess, paramType);
     } else {
@@ -74,6 +77,38 @@ public class ToJsonGenerator extends ConversionGenerator {
               varName,
               jsonValueExpression(getRawClass(paramType), valueAccess)));
     }
+  }
+
+  private void generateEither3Conversion(
+      String indent, PrintWriter out, String varName, String valueAccess, Type paramType) {
+    String innerName = varName + "e";
+
+    out.println(indent + String.format("%1$s %2$s;", json.element(), varName));
+    out.println(indent + String.format("if (%1$s.isFirst()) {", valueAccess));
+    generateToJson(
+        indent + INDENT,
+        out,
+        innerName,
+        valueAccess + ".getFirst()",
+        EitherUtil.getFirstDisjointType(paramType));
+    out.println(indent + INDENT + String.format("%1$s= %2$s;", varName, innerName));
+    out.println(indent + String.format("} else if (%1$s.isSecond()) {", valueAccess));
+    generateToJson(
+        indent + INDENT,
+        out,
+        innerName,
+        valueAccess + ".getSecond()",
+        EitherUtil.getSecondDisjointType(paramType));
+    out.println(indent + INDENT + String.format("%1$s= %2$s;", varName, innerName));
+    out.println(indent + "} else  {");
+    generateToJson(
+        indent + INDENT,
+        out,
+        innerName,
+        valueAccess + ".getThird()",
+        EitherUtil.getThirdDisjointType(paramType));
+    out.println(indent + INDENT + String.format("%1$s= %2$s;", varName, innerName));
+    out.println(indent + "}");
   }
 
   private void generateEitherConversion(
@@ -102,6 +137,9 @@ public class ToJsonGenerator extends ConversionGenerator {
 
   private void generateMapConversion(
       String indent, PrintWriter out, String varName, String valueAccess, Type paramType) {
+    if (!(paramType instanceof ParameterizedType)) {
+      paramType = ((Class<?>) paramType).getGenericSuperclass();
+    }
     ParameterizedType genericType = (ParameterizedType) paramType;
     Type valueType = genericType.getActualTypeArguments()[1];
     String containedName = varName + "X";

